File: measuretool/MeasureTools/ELPShape2Equation.m
function parm = ELPShape2Equation(elpshape)
% 形状参数转方程参数：前两个中心点，然后半长短轴
xc = elpshape(1); yc = elpshape(2); a = elpshape(3); b = elpshape(4); theta = elpshape(5);

parm = zeros(1,5);

parm(1) = cos(theta)^2/a^2 + sin(theta)^2/b^2;
parm(2) = -(sin(2*theta)*(a^2 - b^2))/(2*a^2*b^2);
parm(3) = cos(theta)^2/b^2 + sin(theta)^2/a^2;
parm(4) = (-a^2*xc*sin(theta)^2 + (a^2*yc*sin(2*theta))/2)/(a^2*b^2) - (xc*cos(theta)^2 + (yc*sin(2*theta))/2)/a^2;
parm(5) = (-a^2*yc*cos(theta)^2 + (a^2*xc*sin(2*theta))/2)/(a^2*b^2) - ((xc*sin(2*theta))/2 + yc*sin(theta)^2)/a^2;
parm(6) = (xc*cos(theta) + yc*sin(theta))^2/a^2 + (yc*cos(theta) - xc*sin(theta))^2/b^2 - 1;
k = parm(1)*parm(3)-parm(2)^2;
parm = parm/sqrt(abs(k));


end

File: measuretool/MeasureTools/ELPShape2Equation.cpp
#include "MeasureTools.h"


#include <iostream>

using namespace std;

void ELPShape2Equation(double *elpshape, double *outparms)
{
	double xc, yc, a, b, theta;
	xc = elpshape[0], yc = elpshape[1], a = elpshape[2], b = elpshape[3], theta = elpshape[4];

	double parm[6];

	parm[0] = cos(theta)*cos(theta) / (a*a) + pow(sin(theta), 2) / (b*b);
	parm[1] = -(sin(2 * theta)*(a*a - b*b)) / (2 * a*a*b*b);
	parm[2] = pow(cos(theta), 2) / (b*b) + pow(sin(theta), 2) / (a*a);
	parm[3] = (-a*a*xc*pow(sin(theta), 2) + a*a*yc*sin(2 * theta) / 2) / (a*a*b*b) - (xc*pow(cos(theta), 2) + yc*sin(2 * theta) / 2) / (a*a);
	parm[4] = (-a*a*yc*pow(cos(theta), 2) + a*a*xc*sin(2 * theta) / 2) / (a*a*b*b) - (yc*pow(sin(theta), 2) + xc*sin(2 * theta) / 2) / (a*a);
	parm[5] = pow(xc*cos(theta) + yc*sin(theta), 2) / (a*a) + pow(yc*cos(theta) - xc*sin(theta), 2) / (b*b) - 1;

	double k = parm[0] * parm[2] - parm[1] * parm[1];


	for (int i = 0; i < 6; i++)
		outparms[i] = parm[i] / sqrt(abs(k));



}